### PR TITLE
refactor(es6): Update DefinePlugin to es2015

### PR DESCRIPTION
**Pull request by neeharv**
**What kind of change does this PR introduce?**
Refactor

**Did you add tests for your changes?**
Existing tests should cover this refactor

**If relevant, link to documentation update:**

**Summary**

**Does this PR introduce a breaking change?**
Nope

**Other information**

### CODE REVIEW

**Comment by neeharv** [comment]
A few refactor notes, which I guess will be answered when a style guide is published for es2015 code

1. A bunch of code here is of the form `"(" + code.toString() + ")"`, should I switch that over to ES6 template literals?
2. There are a bunch of functions defined inline, should they be switched to `cons fnName = () => {}` to make their scoped nature explicit?

File: lib/DefinePlugin.js [hunk 1]
```diff
@@ -2,129 +2,123 @@
...
+					if(isTypeof) key = key.replace(/^typeof\s+/, "");
+					let recurse = false;
+					let recurseTypeof = false;
+					code = toCode(code);
+					if(!isTypeof) {
+						parser.plugin("can-rename " + key, () => true);
+						parser.plugin("evaluate Identifier " + key, (expr) => {
+							if(recurse) return;
+							recurse = true;
+							let res = this.evaluate(code);
```
Comment (sokra) [inline lib/DefinePlugin.js]: `this` is used wrongly here. Use `parser` instead.

File: lib/DefinePlugin.js [hunk 2]
```diff
@@ -2,129 +2,121 @@
...
-						return true;
+				function applyDefineKey(prefix, key) {
+					const splittedKey = key.split(".");
+					splittedKey.slice(1).forEach((_, i) => {
+						const fullKey = prefix + splittedKey.slice(0, i + 1).join(".");
+						parser.plugin("can-rename " + fullKey, () => true);
 					});
-					parser.plugin("evaluate Identifier " + key, function(expr) {
-						if(recurse) return;
-						recurse = true;
```
Comment (sokra) [inline lib/DefinePlugin.js]: @timse see here. This line was accidentally removed.

Comment (sokra) [inline lib/DefinePlugin.js]: It should allow to define `ABC = ABC`

Comment (timse) [inline lib/DefinePlugin.js]: Aye! Will add this back. What could this lead this to? So i can check if there are open issues related to this/write a test for it

**Pull request merged**